### PR TITLE
[Feature] Support clucene deps

### LINKED ISSUE #34019: Support fulltext search with inverted index
## Feature request
**Motivation**
Inverted index is a commonly used data structure in information retrieval systems, which maps the "word-document" relationship to the "document-word" relationship. In an inverted index, each unique word has an associated index pointing to a list of documents containing that word. This indexing method makes document search based on words highly efficient.

In OLAP engines, inverted indexes and information retrieval play a crucial role. OLAP engines are typically used for handling complex analytical queries, which often involve large amounts of data and require aggregation and filtering across multiple dimensions.

Inverted indexes can greatly improve the performance of these queries. For example, if you want to find all data rows containing a specific word, you can directly look up the inverted index instead of scanning the entire dataset. This can significantly optimize the query scan volume, especially when retrieving a small amount of data from massive datasets, resulting in greatly improved query performance.

Currently, several OLAP engines, such as `Doris` and `ClickHouse`, have their own implementation of inverted indexes.

**Solution**

We want to abstract an implementation framework for inverted indexes, and CLucene is the first specific implementation we are trying.
flowchart (1)

When we use inverted index, we can use grammar below to manage, write and query index.
```
// create table with inverted index
CREATE TABLE table_name
(
  columns_difinition,
  INDEX idx_name1(column_name1) USING GIN [PROPERTIES("analyzer" = "english|chinese|none", ...)] [COMMENT 'your comment']
  INDEX idx_name2(column_name2) USING GIN [PROPERTIES("analyzer" = "english|chinese|none", ...)] [COMMENT 'your comment']
)

// create index for table
-- grammar 1
CREATE INDEX idx_name ON table_name(column_name) USING GIN [PROPERTIES("analyzer" = "english|chinese|none", ...)] [COMMENT 'your comment'];
-- grammar 2
ALTER TABLE table_name ADD INDEX idx_name(column_name) USING GIN [PROPERTIES("analyzer" = "english|chinese|none", ...)] [COMMENT 'your comment'];
-- grammar 3 
ALTER TABLE table_name MODIFY INDEX idx_name(column_name)  [PROPERTIES] [COMMENT 'your comment']

// drop index
-- grammar 1
DROP INDEX idx_name ON table_name;
-- grammar 2
ALTER TABLE table_name DROP INDEX idx_name;

// query match inverted index with basic binary comparison operators: =,!=,>,>=,<,<=
select * from table_name where column='xxx'; // match

// furthermore, we can also use the following functions to have complete full-text search ability.
// terms query
select * from table where match_terms(f, ['n1', 'n2']);

// phrase query
select * from table where match_phrase(f1, "n1", "n2", 2);

// wildcard query
select * from table where match_wildcard(f1, "*15*");

// fuzzy query
select * from table where match_fuzzy(f1, "thinkbad", 1);

// range query
select * from table where match_range(f1, 1, 10, false, true);

// boost for query
select * from table where boost(terms_query(name, ['n1']), 10) and [other conditions];
```

### PR DESCRIPTION
**Pull request by dujijun007**
Fixes #34019

## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [ ] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: be/CMakeLists.txt [hunk 1]
```diff
@@ -174,6 +174,17 @@ if (${JEMALLOC_DEBUG})
...
+set_target_properties(ic PROPERTIES IMPORTED_LOCATION ${THIRDPARTY_DIR}/lib/libic.a)
+
+add_library(clucene-core STATIC IMPORTED)
+set_target_properties(clucene-core PROPERTIES IMPORTED_LOCATION ${THIRDPARTY_DIR}/lib/libclucene-core-static.a)
+
+add_library(clucene-shared STATIC IMPORTED)
+set_target_properties(clucene-shared PROPERTIES IMPORTED_LOCATION ${THIRDPARTY_DIR}/lib/libclucene-shared-static.a)
+
+add_library(clucene-contribs-lib STATIC IMPORTED)
+set_target_properties(clucene-contribs-lib PROPERTIES IMPORTED_LOCATION ${THIRDPARTY_DIR}/lib/libclucene-contribs-lib.a)
```
Comment (imay) [inline be/CMakeLists.txt:187]: You need to change these codes in the following PR.
In this PR, only thirdparty should be changed

File: thirdparty/build-thirdparty.sh [hunk 2]
```diff
@@ -1205,6 +1205,51 @@ build_libdeflate() {
...
+    cd "$TP_SOURCE_DIR/${CLUCENE_SOURCE}"
+
+    mkdir -p "${BUILD_DIR}"
+    cd "${BUILD_DIR}"
+    rm -rf CMakeCache.txt CMakeFiles/
+
+    ${CMAKE_CMD} -G "${CMAKE_GENERATOR}" \
+        -DCMAKE_INSTALL_PREFIX="$TP_INSTALL_DIR" \
+        -DBUILD_STATIC_LIBRARIES=ON \
+        -DBUILD_SHARED_LIBRARIES=OFF \
```
Comment (imay) [inline thirdparty/build-thirdparty.sh:1234]: `-DCMAKE_INSTALL_LIBDIR=lib64`
add this to make sure the library will be generated in the same path in all platforms

File: thirdparty/vars.sh [hunk 3]
```diff
@@ -409,8 +409,14 @@ LLVM_NAME="llvm-project-16.0.6.src.tar.xz"
 LLVM_SOURCE="llvm-project-16.0.6.src"
 LLVM_MD5SUM="dc13938a604f70379d3b38d09031de98"
 
+#clucene
+CLUCENE_DOWNLOAD="https://raw.githubusercontent.com/dujijun007/clucene/main/libclucene-v2.4.4.zip"
```
Comment (imay) [inline thirdparty/vars.sh:413]: Why not use the official download path?

File: thirdparty/build-thirdparty.sh (hunk 2 above)
Comment (dujijun007) [reply thirdparty/build-thirdparty.sh:1235]: done

File: be/CMakeLists.txt (hunk 1 above)
Comment (dujijun007) [reply be/CMakeLists.txt]: OK, I will remove them and change this in the next PR.

File: thirdparty/vars.sh (hunk 3 above)
Comment (dujijun007) [reply thirdparty/vars.sh]: I used the archive url to replace this

File: build.sh [hunk 4]
```diff
@@ -435,6 +435,8 @@ if [ ${BUILD_FE} -eq 1 -o ${BUILD_SPARK_DPP} -eq 1 ]; then
         cp -r -p ${STARROCKS_HOME}/fe/spark-dpp/target/spark-dpp-*-jar-with-dependencies.jar ${STARROCKS_OUTPUT}/fe/spark-dpp/
         cp -r -p ${STARROCKS_THIRDPARTY}/installed/jindosdk/* ${STARROCKS_OUTPUT}/fe/lib/
         cp -r -p ${STARROCKS_THIRDPARTY}/installed/async-profiler/* ${STARROCKS_OUTPUT}/fe/bin/
+        cp -r -p ${STARROCKS_THIRDPARTY}/installed/share/dict ${STARROCKS_OUTPUT}/be/
```
Comment (imay) [inline build.sh:438]: This can be introduced in the next PR too.

Comment (dujijun007) [reply build.sh]: Got it

File: thirdparty/vars.sh [hunk 5]
```diff
@@ -409,8 +409,14 @@ LLVM_NAME="llvm-project-16.0.6.src.tar.xz"
 LLVM_SOURCE="llvm-project-16.0.6.src"
 LLVM_MD5SUM="dc13938a604f70379d3b38d09031de98"
 
+#clucene
+CLUCENE_DOWNLOAD="https://github.com/dujijun007/clucene/archive/refs/tags/2.4.4/libclucene-2.4.4.zip"
```
Comment (kevincai) [inline thirdparty/vars.sh:413]: any customization to the original source code? Is it possible to use its origin sourceforge download link?

Comment (dujijun007) [reply thirdparty/vars.sh:413]: @kevincai There are some bugfix commits, how about uploading to the thirdparty repo of StarRocks, in order to do some modifying in the future.

Comment (kevincai) [reply thirdparty/vars.sh:413]: yes, we can create a repo under starrocks/ with exactly the same commit from upstream, and then apply necessary bugfixes.

Comment (kevincai) [reply thirdparty/vars.sh:413]: please point me the base commit you are referring in your own repo, I can help to create one under starrocks org.

Comment (dujijun007) [reply thirdparty/vars.sh:413]: You can init from [this commit](https://github.com/dujijun007/clucene/commit/7e80d5bdf6bbe4a6321d488dff455e063f652484)

Comment (kevincai) [reply thirdparty/vars.sh:413]: I have created https://github.com/StarRocks/clucene and preserves all the commit history from clucene sourceforge git repo. Can you point me which commit from the original upstream, we can branch from the commit and apply incremental fixes.

Comment (dujijun007) [reply thirdparty/vars.sh:413]: Thanks for your help, can you publish a raw release with the tar file in this repo? Then I can get the right clucene dependency url.

Comment (kevincai) [reply thirdparty/vars.sh:413]: I have made a tag pointed to the main latest commit, you can use the link: https://github.com/StarRocks/clucene/archive/refs/tags/starrocks-2023.11.06.tar.gz

File: thirdparty/vars.sh [hunk 6]
```diff
@@ -409,8 +409,14 @@ LLVM_NAME="llvm-project-16.0.6.src.tar.xz"
 LLVM_SOURCE="llvm-project-16.0.6.src"
 LLVM_MD5SUM="dc13938a604f70379d3b38d09031de98"
 
+#clucene
+CLUCENE_DOWNLOAD="https://github.com/StarRocks/clucene/archive/refs/tags/starrocks-2023.11.06.tar.gz"
+CLUCENE_NAME="libclucene-v2.4.4.zip"
```
Comment (kevincai) [inline thirdparty/vars.sh:414]: these names may need update as well.

Comment (dujijun007) [reply thirdparty/vars.sh]: Please check again

File: thirdparty/build-thirdparty.sh [hunk 7]
```diff
@@ -1205,6 +1205,51 @@ build_libdeflate() {
...
+        -DCMAKE_INSTALL_PREFIX="$TP_INSTALL_DIR" \
+        -DCMAKE_INSTALL_LIBDIR=lib64 \
+        -DBUILD_STATIC_LIBRARIES=ON \
+        -DBUILD_SHARED_LIBRARIES=OFF \
+        -DBOOST_ROOT="$TP_INSTALL_DIR" \
+        -DZLIB_ROOT="$TP_INSTALL_DIR" \
+        -DCMAKE_CXX_FLAGS="-g -fno-omit-frame-pointer -Wno-narrowing" \
+        -DUSE_STAT64=0 \
+        -DUSE_AVX2="${USE_AVX2}" \
+        -DUSE_BTHREAD="${USE_BTHREAD}" \
```
Comment (kevincai) [inline thirdparty/build-thirdparty.sh:1241]: the upstream version doesn't support these two options

File: thirdparty/build-thirdparty.sh [hunk 8]
```diff
@@ -1205,6 +1205,51 @@ build_libdeflate() {
     ${BUILD_SYSTEM} install
 }
 
+#clucene
+build_clucene() {
+    if [[ "$(uname -m)" == 'x86_64' ]]; then
+        USE_AVX2="${USE_AVX2:-1}"
```
Comment (kevincai) [inline thirdparty/build-thirdparty.sh:1211]: not supported.

File: thirdparty/build-thirdparty.sh (hunk 7 above)
Comment (dujijun007) [reply thirdparty/build-thirdparty.sh:1241]: I will remove these compile options first, then upgrade clucene to support them in following PR.

File: thirdparty/build-thirdparty.sh [hunk 9]
```diff
@@ -1205,6 +1205,36 @@ build_libdeflate() {
...
+        -DBUILD_STATIC_LIBRARIES=ON \
+        -DBUILD_SHARED_LIBRARIES=OFF \
+        -DBOOST_ROOT="$TP_INSTALL_DIR" \
+        -DZLIB_ROOT="$TP_INSTALL_DIR" \
+        -DCMAKE_CXX_FLAGS="-g -fno-omit-frame-pointer -Wno-narrowing" \
+        -DUSE_STAT64=0 \
+        -DCMAKE_BUILD_TYPE=Release \
+        -DBUILD_CONTRIBS_LIB=ON ..
+    ${BUILD_SYSTEM} -j "${PARALLEL}"
+    ${BUILD_SYSTEM} install
```
Comment (kevincai) [inline thirdparty/build-thirdparty.sh:1230]: this install will include additional boost headers, a little worried about this, will it cause boost header conflict?

**Review by kevincai — APPROVED** [review]

File: thirdparty/build-thirdparty.sh (hunk 9 above)
Comment (dujijun007) [reply thirdparty/build-thirdparty.sh:1230]: I did not meet any conflict when I compiled and tested, but your worry makes sense, I will keep an eye on this issue and try to address it in the subsequent refactoring.

**Review by imay — APPROVED** [review]

**Pull request merged**